enhancements/network/external-lb-vips.md

```diff
@@ -0,0 +1,213 @@
+  remain the HAproxy/Keepalived/CoreDNS stack.
+* Supporting a mix of both (internal + external loadbalancers for the API & Ingress VIPs), customers
+  will have to choose a scenario.
+
+## Proposal
+
+We will make some change in `install-config.yaml` so customer can configure the
+deployment to use an external loadbalancer if they want.
+Then the components which interact with networking will have to check this
+new field and handle this scenario appropriately.
+
+### Workflow Description
+
+1. The customer will configure their external loadbalancer(s), which includes configuring the VIPs
+   for API and Ingress.
+2. The customer will populate their external DNS server with required records for api, api-int, and ingress.
+3. The customer will write the `install-config.yaml` file and set the new `loadBalancerType` field to `external`, and
+   will not set the fields for apiVIP and ingressVIP. We should fail if they set these mutually exclusive parameters.
+4. In the installer, if the platform does support external loadbalancer, some validation will be run
+   to check that the VIPs can already be resolved (the validation already exists for vsphere, we need to
+    make it re-usable). A new field in `PlatformStatus` will be set.
+5. In machine-config-operator, if LB type is external, the internal loadbalancer will not be
+   deployed.
+6. When using an external loadbalancer, the VIPs fields won't be set in `PlatformStatus` and the
+   OpenShift components won't use it when they're not present and LB type is external.
+7. After the deployment, the IP address of a service (e.g. API) can change, as long as the external
+   loadbalancer is updated and the DNS record has been updated and the results are propagated as well.
+
+### API Extensions
+
+A new field will be added to the `PlatformStatus` of the infrastructure object to tell
+the kind of the loadbalancer.
+
+```golang
+// LoadBalancerType can be either external or internal 
+type LoadBalancerType string
+const (
+	// LoadBalancerTypeExternal
+	LoadBalancerTypeExternal LoadBalancerType = "external"
+	// LoadBalancerTypeInternal
+	LoadBalancerTypeInternal LoadBalancerType = "internal"
+)
+// VSpherePlatformStatus holds the current status of the vSphere infrastructure provider.
+type VSpherePlatformStatus struct {
+  // loadBalancerType defines if a internal or external load balancer
+  // is to be used for the openshift api and ingress endpoints.
+  // When set to Internal the static pods defined in machine config operator
+  // will be used.
+  // When set to External the static pods will not be deployed. It will be 
+  // expected that the LB and DNS A records be pre-configured prior to installation.
+  // +kubebuilder:default=internal
+  // +kubebuilder:validation:Enum=internal;external
+  LoadBalancer LoadBalancerType `json:"loadBalancerType,omitempty"`
+}
+```
+
+This field will be set in the `install-config.yaml`:
+```yaml
+platform:
+  vsphere:
+    loadBalancerType: external
+```
```

Comment:
e.g.

```yaml
platform:
  loadBalancerType: external
  vsphere:
    ...
```

Reply:
yes, it would be nice indeed (to confirm with @patrickdillon and his team).

question for reviewers: is there any openshift/api impact on moving `loadBalancerType` out of `platform.<platform name>.loadBalancerType` into `platform.loadBalancerType` ? Or is it just an installer thing?

Reply:
So I checked in the installer and there is no parameter outside of a <platform> in `platform` section of install-config.yaml.

I'm sure Patrick has a great reasoning on why.

Reply:
The `platform` is a discriminated union, so it should contain only the discriminator and the union members. Additional fields like `loadBalancerType` would then add to confusion, it would need to be outside of platform

Reply:
Fields in `platform` should be specific to the platform/provider. If `LoadBalancer`applies (or could potentially apply) to all platforms then it should be at the top-level--not nested under `platform`. 

[`PublishStrategy`](https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L150) in the install config is a decent example of this.